pkg/storage/plan_storage.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/abcxyz/guardian/pkg/github"
+	"github.com/abcxyz/guardian/pkg/platform"
+)
+
+var _ PlanStorage = (*PlanStorageClient)(nil)
+
+// PlanStorageClient implements [PlanStorage].
+type PlanStorageClient struct {
```

Comment:
I'm confused by this encapsulation (and the overload of the word "storage"). I feel like the client should be constructed outside and passed in as a parameter instead.

Reply:
to me the storage flag is "owned" by the plan/apply command. it should be used to parse a storage type and generate a client.

separately we need configs from the platform to generate storage urls (e.g. the path prefix for uniqueness when using github and pull requests).

does it make sense to add a helper function to the platform config to generate the required "plan storage configs", accepting the storage flag value?

Reply:
> to me the storage flag is "owned" by the plan/apply command

I think that's correct, but why is it generating a specific client for Plan vs Apply? It should be creating a generic "storage" interface (CREATE, READ, DELETE) for gcs and local. Then in the "storage" package, we instantiate the storage client from the flag inputs.

Reply:
updated per offline discussions